### ISSUE TITLE
Questions about training and training stability

### ISSUE THREAD

**Issue by greeneggsandyaml** [madebyollin/taesd#11]
Hello,

Thank you so much for all your awesome work. It is really great stuff! 

I have some questions about training (not answered by #2). First, if it is possible, would you be able to release your training code? It would be super helpful. 

I'm asking because I'm looking to train some autoencoders that combine multiple types of image information (i.e. rgb, depth, segmentation, etc.). I have a training script at the moment (based on diffusers [here](https://github.com/huggingface/diffusers/issues/3726)), but I'm finding that training is very unstable. I'm constantly getting NaNs. Did you find that any tricks were necessary to finetune the VAE without getting NaNs? What learning rate / batch size did you use? 

Also, if you have any more experiments that you want to run, I have access to some good resources (8 A100s), so let me know. I feel like these autoencoders are quite under-explored and a lot more (semantic and geometric) information could be integrated into them.

Best,
greeneggsandyaml

**Comment by madebyollin** [comment]
If I can get VAE training code that works reliably and isn't a mess to read, I'll definitely release it - still experimenting with various approaches for that though...

For the TAESD weights in the repo, my optimizer was `th.optim.Adam(model.parameters(), 3e-4, betas=(0.9, 0.9), amsgrad=True)` and batch size 16 - nothing fancy (trained on 1xA10 :P). My adversarial loss was "relativistic" (penalizing `distance(disc(real).mean(), disc(fake).mean())`, rather than just `disc(fake).mean()`) and I used a replay buffer of fakes for the discriminator training, both of which may have helped with stability. I also used several auxiliary losses (LPIPS + frequency-amplitude matching + MSE at 1/8 res) for the most recent model, which helped reduce the dependence on the adversarial loss a bit. I don't remember any persistent instability issues with this setup.

**Comment by madebyollin** [comment]
BTW, it looks like the MosaicML folks are working on a cleaned up version of the Latent Diffusion VAE training code here https://github.com/mosaicml/diffusion/pull/79

**Comment by madebyollin** [comment]
Adding some more info here.

## Changes in 1.2

For TAESD 1.2, I removed the LPIPS and other icky hand-coded losses (now just using adversarial + very faint lowres MSE). I also added adversarial loss to the encoder training as well (though I'm not sure it made a difference).

## Various questions I've seen

1. **Are the decoder targets GT images or SD-decoded images?** GT; TAESD's decoder is a standalone conditional GAN, not a distilled model.
2. **What dataset was used?** Depends on model version, but usually some mix of photos (e.g. [laion-aesthetic](https://laion.ai/blog/laion-aesthetics/)) and illustrations (e.g. [danbooru2021](https://gwern.net/danbooru2021#samples)), with some color / geometric augmentations
3. **Do you delay adversarial loss until a certain number of steps (like the SD VAE does)?** I usually prefer to start from a pretrained decoder model, but I don't have some specific number of steps in mind.
4. **What do you mean by low-res MSE loss?** like `F.mse_loss(F.avg_pool2d(decoded, 8), F.avg_pool2d(real_images, 8))`. Just making sure that the color of each 8x8 patch is approximately correct.
5. **Which reference VAEs do you use?** https://huggingface.co/stabilityai/sd-vae-ft-ema and https://huggingface.co/madebyollin/sdxl-vae-fp16-fix - these are used to supervise the encoder and also as a gold standard for decoder quality.

## Various figures

### Color Augmentation

Color augmentation (occasional hue / saturation shifting of input images) helped improve reproduction of saturated colors (which are otherwise rare in aesthetic datasets)

<img width="512" alt="image" src="https://github.com/madebyollin/taesd/assets/8733615/ad8c6c61-65ad-4832-be62-c1c43a16b605">

### Downsides of different losses

MSE/MAE can make everything very smooth (top is GT, bottom is a simple MSE-only decoder)
<img width="256" alt="image" src="https://github.com/madebyollin/taesd/assets/8733615/8a964b72-2008-487d-8c37-17e11fa5966e">

LPIPS can cause recognizable artifacts on faces & eyes (top is from a run with LPIPS, bottom is a run without it)

<img width="191" alt="image" src="https://github.com/madebyollin/taesd/assets/8733615/c98a663d-0428-4f15-9265-7a141e221adb">

Adversarial loss can cause divergence if not handled properly:

<img width="512" alt="image" src="https://github.com/madebyollin/taesd/assets/8733615/9a4bcd7f-22af-4253-8403-2ba52a477ca9">

### Blue eyes

I don't remember what caused this

<img width="256" alt="image" src="https://github.com/madebyollin/taesd/assets/8733615/710f2527-b666-4eae-9a1d-5ba2d1609b1f">

**Comment by bcw222** [comment]
> If I can get VAE training code that works reliably and isn't a mess to read, I'll definitely release it - still experimenting with various approaches for that though...

Hello Ollin, I'm a Machine Learning beginner and want to train a TAESD (maybe a modified version) from scratch.
May you release the training code (even an early version) now?

**Comment by madebyollin** [comment]
@bcw222 I posted https://github.com/madebyollin/seraena/blob/main/TAESDXL_Training_Example.ipynb which should work as a starting point (and it does most of the complicated adversarial loss part). You can try adding additional pooled-MSE / LPIPS losses to speed up convergence

**Comment by madebyollin** [comment]
@app-phtu The training example notebook only uses 96x96 images, which iirc end up "more-saturated" than usual (the SD-VAE has some resolution-dependent behavior that the TAESD decoders can't mimic). I would recommend finetuning on 256x256 images (the SD VAE training resolution) to calibrate the tiny encoder/decoder color saturation for high-resolution images.

Beyond that, for encoder losses, I have used the following for various different tiny VAEs:
1. MSE loss (baseline)
2. L1 loss (generally "sharper" than MSE, albeit slightly)
3. Additional feature-matching loss from the early decoder features (e.g. `F.l1_loss(swa_version_of_tiny_decoder[:6](real_latents), swa_version_of_tiny_decoder[:6](fake_latents)`)

For the TAESDXL encoder I think I just used L1 encoder loss at 256x256. So I would suggest trying that, and then adding in feature-matching if L1 is still not enough.

**Comment by app-phtu** [comment]
Hi @madebyollin, thanks so much for your insights and quick replies. This is very helpful! I find I have to add early layers of teacher decoder (not the distilled tiny decoder because I only want distilled encoder) features. But when I add the decoder features, I find I get blob artifacts near the image corners and they are fixed in one spot across all images. Have you experienced similar issues? Thank a lot!

<img width="59" height="60" alt="Image" src="https://github.com/user-attachments/assets/8d2269bf-1aee-40b0-a7fe-e8dd1d41c610" />

<img width="58" height="52" alt="Image" src="https://github.com/user-attachments/assets/51f41afb-43be-4c5d-9359-2fd560290f07" />

**Comment by madebyollin** [comment]
@app-phtu Hmm I have only tried featurematching with the distilled tiny decoder features, which didn't have that issue. I've seen similar-looking blob artifacts show up when the model is fighting against normalization/attention layers (see [StyleGAN 2](https://openaccess.thecvf.com/content_CVPR_2020/papers/Karras_Analyzing_and_Improving_the_Image_Quality_of_StyleGAN_CVPR_2020_paper.pdf) figure 1) but for tiny VAEs I'm not sure how normalization layers would get involved 🤔.

I would suggest first making sure that running the full decoder actually decodes valid, in-distribution-looking images for both real/fake latents (if not, matching the intermediate features probably won't work well). You could also try visualizing the forward/backward pass intermediate tensors to narrow down where the blob artifact starts showing up.
